### ISSUE TITLE
Allow connecting to self-hosted instances with self-signed certs

### ISSUE THREAD

**Issue by randrej** [wger-project/flutter#116]
Currently, the app doesn't connect to my local instance because it has a self-signed cert, even though I've imported the CA in Android.

Can you add an option to allow logging with local cer (or an option to disable cert verification) when you select local server?

**Comment by rolandgeider** [comment]
This is definitely something we didn't test. It seems flutter's HTTP client uses the list of CAs from mozilla, but it's possible to add own ones

https://api.flutter.dev/flutter/dart-io/HttpClient-class.html

**Comment by randrej** [comment]
Thanks for the reply.

I seem to have found a way to fix this, but I'm not very familiar with Flutter, being primarily a C++/C programmer.

I found in a StackOverflow post that you can replace the handler for bad certificates in your HttpClient() object: https://stackoverflow.com/a/54359013/1307661

You could add a setting that enables this fix, exposed through a checkbox (called 'self-signed certificate') on the login screen (when you pick an alternative server).
Then, in any part of the program where you instantiate a `http.Client`, replace it with either:

- a subclass that has the bad cert handler set according to the proposed setting
- a call to a factory function that produces  `http.Client`s with the bad cert handler set according to the proposed setting

**Comment by kovdan01** [comment]
Hello @rolandgeider, found a PR on the same topic which might be helpful for you: https://github.com/libre-tube/LibreTube/pull/806. Tested that and it works. TLDR: you just need to alter android manifest appropriately. A drawback of such approach is that you decide that your app trusts user-installed certificates at compile time, and you can't change this behavior in runtime. If it's applicable - looks like changing manifest is the simplest way of implementing desired functionality.

**Comment by rolandgeider** [comment]
That sounds great. Just editing the XML would obviously be the fastest solution and I think always trusting user certificates would also be acceptable

**Comment by kovdan01** [comment]
Hello @rolandgeider, please let me know if some work is planned on the issue and if I can help somehow (unfortunately, I'm not a flutter developer and can't implement the feature myself). Thanks!

**Comment by rolandgeider** [comment]
Hi @kovdan01 yes, this is definitely planned. After I finish with the exercise crowd sourcing feature, I'll do this

**Comment by kovdan01** [comment]
Thanks!

**Comment by rolandgeider** [comment]
this took a bit longer, but I have just added the XML allowing user certificates

**Comment by kovdan01** [comment]
Thanks! Looking forward for the next release including the feature

**Comment by rolandgeider** [comment]
it's building right now :)

**Comment by kovdan01** [comment]
Hello @rolandgeider . Thanks for implementing the feature, but it seems that it does not work as intended. It looks like that you forgot to add a line in AndroidManifest.xml describing that a network_security_config.xml is present (which was introduced in 9db063d65c055be72703f70874e4bf02710f2ce3). At least, I've tested the new version and it behaves the same in terms of connecting to instances with self-signed certs. So, the issue will be possibly resolved after altering AndroidManifest.xml properly. See, for example, https://github.com/Bnyro/LibreTube/blob/aadf6d441a498ac9866c395d29a1da4df1a62fe9/app/src/main/AndroidManifest.xml#L19.

**Comment by rolandgeider** [comment]
oh no 😅

**Comment by rolandgeider** [comment]
New release is being reviewed by the Play Store Gods

**Comment by vayan** [comment]
is this live now? can't find any setting

**Comment by kovdan01** [comment]
@vayan 
> is this live now? can't find any setting

I might be mistaken, but this is intended to work without any settings, just by altering the manifest in order to allow user certificates.

The last time I checked the issue was still there, but it looks like it's flutter-related and can't be fixed on the wger side. I found some related issues previously, but unfortunately I'm unable to find them right now to leave links here.

I personally switched to using certificates from Let's Encrypt - they are free and do not cause issues. Self-signed ones need to be supported by every single app you want to connect from, so I just gave up.

P.S. Don't take this as an "official" answer, I'm not related to wger development.

**Comment by rolandgeider** [comment]
Exactly, this is a flag set during compilation. I haven't been able to test it because I don't have the appropriate setup. If there is anything else that needs to be done from our side, we'll try to fix it, but yeah in general using let's encrypt is much easier and you will avoid all the pain that comes from self signed certificates.

**Comment by vayan** [comment]
Got it! thanks :) yeah my instance is on a private tailscale network so I can't get a let's encrypt cert. 

but it's ok I'll switch to http for the mobile app! Just wanted to know if I missed something 😄 

Thanks for the great work!

**Comment by rolandgeider** [comment]
Closing the issue now, if there any issues, feel free to repoen

**Comment by GameRebellStudio** [comment]
Hi guys! I came accross this post when I was trying to connect the Android application to my newly setup server. It appears that there's still an issue with self-signed/untrusted certificates. I use certificates signed by local CA, since it's not viable for me to use Let's Encrypt certificates due to the nature of my environment.

I get the following error when trying to login with HTTPS. Is there anything I can do about this? Thank you very much :)

<img src="https://github.com/wger-project/flutter/assets/14983333/0234a0c4-229d-48de-ade1-64a195a69ad8" width="200">
<img src="https://github.com/wger-project/flutter/assets/14983333/48debcf0-901a-4ed5-8d63-5b05cfdd6602" width="200">

**Comment by GameRebellStudio** [comment]
@rolandgeider I'm using the F-Droid release, v1.6.2 if that helps

**Comment by rolandgeider** [comment]
so the flag didn't solve it, damn

**Comment by rolandgeider** [comment]
we've finally merged the PR that allows using self-signed certificates 🥳 

This is a toggle in the start page and will be available in the next version

<img width="446" height="912" alt="Image" src="https://github.com/user-attachments/assets/e12199b4-635d-4bfb-b47b-7a4a1592fcb6" />

**Comment by Romloader** [comment]
> we've finally merged the PR that allows using self-signed certificates 🥳 
> 
> This is a toggle in the start page and will be available in the next version
> 
> <img width="446" height="912" alt="Image" src="https://github.com/user-attachments/assets/e12199b4-635d-4bfb-b47b-7a4a1592fcb6" />

Hey @rolandgeider,

first of all thanks for all the work you and the other contributors do.

After this feature was released, I managed to sign up to my instance with a private CA just fine, so the login works.
However, some other things like syncronization still complain about using a self signed certificate.
Also, for example, the weight tab doesn't load any data.

I don't know if this has to do with the app or with settings on my end.

**Comment by rolandgeider** [comment]
argh, you are right, the powersync service runs in a different isolate and wasn't getting the override. The next version that we're releasing this week with the health sync imports will have the fix.

Third time's the charm 😅

**Comment by Romloader** [comment]
Nice work @rolandgeider !

Synchronization seems to be working on my end now.

The only small thing I see is that the weight data doesn't seem to be loading on the main view of the app (again, if it's a configuration issue on my end then everything is fine)
Update: it also seems that starting a workout shows a warning as the app wasn't able to load the video associated with the current exercise.

**Comment by rolandgeider** [comment]
@Romloader is it only the weight? Note that we recently released a new server version that touched that part of the app. In any case if there's problems there, better open a new issue :)

I'm afraid we will have to ignore the warnings about the videos, those come from the video player and from what I've seen, we can't do much about the certificates there.

**Comment by Romloader** [comment]
@rolandgeider yes you are indeed right.
The issue was on my side: after pulling the latest release there are no issues with the weight data.

Soo everything works (not mentioning videos), thank you so much for this addition ☺️
